contracts/p1/BasketHandler.sol

```diff
@@ -440,12 +460,95 @@ contract BasketHandlerP1 is ComponentP1, IBasketHandler {
 
         for (uint256 i = 0; i < length; ++i) {
             erc20s[i] = address(basket.erc20s[i]);
+            ICollateral coll = assetRegistry.toColl(IERC20(erc20s[i]));
 
             // {qTok} = {tok/BU} * {BU} * {tok} * {qTok/tok}
-            quantities[i] = quantity(basket.erc20s[i]).safeMul(amount, rounding).shiftl_toUint(
-                int8(IERC20Metadata(address(basket.erc20s[i])).decimals()),
-                rounding
+            quantities[i] = _quantity(basket.erc20s[i], coll)
+                .safeMul(amount, rounding)
+                .shiftl_toUint(
+                    int8(IERC20Metadata(address(basket.erc20s[i])).decimals()),
+                    rounding
+                );
+        }
+    }
+
+    /// Return the redemption value of `amount` BUs for a linear combination of historical baskets
+    /// @param basketNonces An array of basket nonces to do redemption from
+    /// @param portions {1} An array of Fix quantities
+    /// @param amount {BU}
+    /// @return erc20s The backing collateral erc20s
+    /// @return quantities {qTok} ERC20 token quantities equal to `amount` BUs
+    // Returns (erc20s, [quantity(e) * amount {as qTok} for e in erc20s])
+    function quoteCustomRedemption(
+        uint48[] memory basketNonces,
```

Comment:
we should write some tests to ensure non-uniqueness cannot be used maliciously. i don't think it can be but it is not immediately obvious

Reply:
Agree we should have a test for it; I'll add it. I do think it's safe, but we should have the test anyway.